phylogeny/prototypeSelection.py

```diff
@@ -369,3 +370,81 @@ def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
                             " number of prototypes could be found."))
 
     return list(prototypes[:num_prototypes])
+
+
+def prototype_selection_constructive_pMedian(dm, num_prototypes):
+    '''Heuristically select k prototypes for given distance matrix.
+
+       Prototype selection is NP-hard. This is an implementation of a greedy
+       correctness heuristic from [1]. The problem is stated as "Given a finite
+       number of users, whose demands for some service are known and must be
+       satisfied, and given a finite set of possible locations among which k
+       must be chosen for the location of service centers, select the locations
+       in such a way as to minimize the total distance travelled by the users."
+       Users are the elements in the distance matrix, k is the number of proto-
+       types. This approach has been termed as the "p-median model".
+
+    Parameters
+    ----------
+    dm: skbio.stats.distance.DistanceMatrix
+        Pairwise distances for all elements in the full set S.
+    num_prototypes: int
+        Number of prototypes to select for distance matrix.
+        Must be >= 2, since a single prototype is useless.
+        Must be smaller than the number of elements in the distance matrix,
+        otherwise no reduction is necessary.
+
+    Returns
+    -------
+    list of str
+        A sequence holding selected prototypes, i.e. a sub-set of the
+        elements in the distance matrix.
+
+    Raises
+    ------
+    ValueError
+        The number of prototypes to be found should be at least 2 and at most
+        one element smaller than elements in the distance matrix. Otherwise, a
+        ValueError is raised.
+
+    Notes
+    -----
+    Timing: %timeit -n 100 prototype_selection_constructive_protoclass(dm, 100)
+            10 loops, best of 3: 32.8 s per loop
+            where the dm holds 27,398 elements
+    function signature with type annotation for future use with python >= 3.5:
+    def prototype_selection_constructive_protoclass(dm: DistanceMatrix,
+    num_prototypes: int) -> List[str]:
+
+    [1] Desire L. Massart, Frank Plastria and Leonard Kaufman.
+        "Non-hierarchical clustering with MASLOC"
+        Pattern Recognition, 1983, Vol. 16, No. 5, pp. 507-516
+    '''
+    if num_prototypes < 2:
+        raise ValueError(("'num_prototypes' must be >= 2, since a single "
+                          "prototype is useless."))
+    if num_prototypes >= len(dm.ids):
+        raise ValueError(("'num_prototypes' must be smaller than the number of"
+                          " elements in the distance matrix, otherwise no "
+                          "reduction is necessary."))
```

Comment:
lines 423: if num_prototypes < 2: in combination with line 426 if num_prototypes >= len(dm.ids):